### ISSUE TITLE
my struct cannot marshal

### ISSUE THREAD

**Issue by ccpwcn** [redis/go-redis#2512]
my struct:
```golang
package model

import (
	jsoniter "github.com/json-iterator/go"
	"gorm.io/gorm"
)

type LoginUserInfo struct {
	UserId   int64
	Username string
}

func (s *LoginUserInfo) MarshalBinary() (data []byte, err error) {
	return jsoniter.Marshal(s)
}
func (s *LoginUserInfo) UnmarshalBinary(data []byte) error {
	return jsoniter.Unmarshal(data, &s)
}
```

use:
```golang
future := time.Duration(3600) * time.Second
mu := model.LoginUserInfo{
	UserId:   row.UserId,
	Username: row.Username,
}
if cmd = redisClient.Set(c, my_consts.LoginUserKey+tokenKey, mu, future); cmd.Err() != nil {
	fmt.Println("保存登录信息失败：" + cmd.Err().Error())
	return
}
fmt.Println("user has been saved to redis")
```

## Expected Behavior

user has been saved to redis

## Current Behavior

保存登录信息失败：redis: can't marshal model.LoginUserInfo (implement encoding.BinaryMarshaler)

## Possible Solution

change the model declaration 【attation: receiver is changed to normal instance, not pointer instance】:
```golang
package model

import (
	jsoniter "github.com/json-iterator/go"
	"gorm.io/gorm"
)

type LoginUserInfo struct {
	UserId   int64
	Username string
}

func (s LoginUserInfo) MarshalBinary() (data []byte, err error) {
	return jsoniter.Marshal(s)
}
func (s *LoginUserInfo) UnmarshalBinary(data []byte) error {
	return jsoniter.Unmarshal(data, s)
}
```

NOW, user has been saved to redis.

## Steps to Reproduce

1. my struct receiver is generic struct instance, it's can be marshal and save into redis
2. my struct receiver is pointer reference struct instance, it's cannot be marshal and save into redis

## Context (Environment)

github.com/redis/go-redis/v9 v9.0.2

## Detailed Description

when my struct as normal instance (not pointer receiver) implement encoding.BinaryMarshaler, I can marshaler the struct instance into redis.

when my struct as instance (is a pointer receiver) implement encoding.BinaryMarshaler, I can't marshaler the struct instance into redis.

WHY???

**Comment by SoulPancake** [comment]
Hi @ccpwcn 
In your struct implementation, you're using github.com/json-iterator/go instead of the standard encoding/json package.
package model

```
import (
	"encoding/json"
	"gorm.io/gorm"
)

type LoginUserInfo struct {
	UserId   int64
	Username string
}

func (s *LoginUserInfo) MarshalBinary() (data []byte, err error) {
	return json.Marshal(s)
}
func (s *LoginUserInfo) UnmarshalBinary(data []byte) error {
	return json.Unmarshal(data, &s)
}
```

**Comment by ccpwcn** [comment]
According to what you said, I did the test again and the error still exists. It's really a sad story.

This is my code:
![image](https://user-images.githubusercontent.com/5465756/229059007-5f30e1ae-e0fc-4e62-8979-68eb999566ef.png)

This is Apifox test the api function:
![image](https://user-images.githubusercontent.com/5465756/229059182-2d7b4bbb-c6c5-4f11-9b56-c20b27462452.png)

**Comment by ccpwcn** [comment]
encoding/json and github.com/json-iterator/go, Their behavior is exactly the same.

**Comment by monkey92t** [comment]
Yes, it was our mistake. I have fixed it in #2489 and we expect to release a new version v9.0.3 over the weekend.

**Comment by ccpwcn** [comment]
@monkey92t Is this bug fixed in v9.0.3?
Today, I update my go.mod dependies, `github.com/redis/go-redis/v9 v9.0.3`
My code still looks like this:
![image](https://user-images.githubusercontent.com/5465756/229393351-a7fc2324-61f9-4b72-b41e-e9f80db715ce.png)

Unfortunately, I still receive the following error message:
`redis: can't marshal model.LoginUserInfo (implement encoding.BinaryMarshaler)"`

One extra point, when I use struct instance, not pointer struct instance, I can marshal it into redis, but I cannot unmarshal it out into memory.

Of course, when `MarshalBinary` function binding into struct instance, `UnmarshalBinary` binding into pointer struct instance. The problems we are talking about will no longer exist:
![image](https://user-images.githubusercontent.com/5465756/229394486-26e96e54-eda0-45de-a35c-250482076958.png)

It was just not in line with our usual code writing habits, and the IDE kept telling me that there was something wrong with my code, which was very uncomfortable for a programmer who wanted to write good code.

In short, mistakes still seem to exist.

**Comment by monkey92t** [comment]
@ccpwcn  Your issue is unrelated to go-redis. You should understand the difference between struct pointers and non-pointers.

I hope the following example is helpful to you.

```go
type Look interface {
	Up() error
}

type My struct{}

func (m *My) Up() error {
	return nil
}

func main() {
	var a any = My{}
	var b any = &My{}

	switch a.(type) {
	case Look:
		fmt.Println("a true")
	}
	switch b.(type) {
	case Look:
		fmt.Println("b true")
	}
}
```

**Comment by ccpwcn** [comment]
@monkey92t Yes, This is my mistakes, Thank you for your tips and help. Thank you very much.🍕🍕🍕

**Comment by hexxv** [comment]
hhhh
